### PR TITLE
Ensure gcs bucket for k8s-infra prow instance

### PR DESCRIPTION
**Pull request by ameukam**
Following prow [documentation](https://github.com/kubernetes/test-infra/blob/master/prow/getting_started_deploy.md#configure-a-gcs-buckethttps://github.com/kubernetes/test-infra/blob/master/prow/getting_started_deploy.md#configure-a-gcs-bucket) guidance :
Create a GCS bucket for tide history and build logs.
Create a service account and grant admin access to the bucket.
Create a service account key and add the generated key to Secret
Manager.

Signed-off-by: Arnaud Meukam <ameukam@gmail.com>

### CODE REVIEW

**Comment by ameukam** [comment]
/retitle Ensure gcs bucket for k8s-infra prow instance

**Review by cpanato — APPROVED** [review]
/lgtm

File: infra/gcp/ensure-main-project.sh
Comment (spiffxp) [inline infra/gcp/ensure-main-project.sh:62]: Assuming this is supposed to be the community equivalent of `kubernetes-jenkins`?
```suggestion
    k8s-infra-prow-results
```

Naming is hard when I overthink things.
- First thought: `k8s-infra-prow-logs` to get the consistent `k8s-infra` prefix
- But...
  - "prow logs" sounds like logs of the prow components (plank, sinker, crier, etc)
  - there are more than "logs" that end up `kubernetes-jenkins`, there are also "artifacts" like junit.xml files, coverage files, metadata, etc.
- OK, so `k8s-infra-prow-artifacts` for "artifacts" produced by prow, including job results
- But...
  - "artifacts" clashes with the sig-release meaning of "binaries and container images that represent a given release of a given subproject"
  - makes me thing it's going to be the binaries/images for prow itself, a.k.a. the equivalent to `gcr.io/k8s-prow`
- OK, so how about just `k8s-infra-prow` and be done with it
- But...
  - If we're doing things over, one of the problems with `kubernetes-jenkins` is that it allows per-object ACLs. They allowed that single bucket to be used for multiple purposes, including occasionally storing secrets or credentials.  
  - We don't want per-object ACLs in kubernetes.io, as they are painful and laborious to maintain, audit and enforce.  We are planning on setting an org policy to explicitly require per-bucket IAM only (aka uniform bucket level access).  
  - Thus we should name this bucket for the main purpose everyone expects of `kubernetes-jenkins`, public read access to things about and created by prow jobs for the kubernetes project.  If other purposes arise, they should result in other buckets named/IAM'ed as such.
- OK, so how about `k8s-infra-prow-results`, as in the "results" of prow job runs
  - I looked around in prow's codebase and docs and couldn't find a consistent term for "things that are uploaded to GCS or S3 by the pod-utilities `sidecar` component and `crier`".  Results isn't just "test results" but from the concept of running a prow job and then looking at the "resulting" files in GCS/S3.
  - I considered "data" but that sounds like "training data" or other metadata that is loaded in to help run a job, where I believe this bucket's intended purpose is storage of things about and created by prow jobs

WDYT about `k8s-infra-prow-results`?

File: infra/gcp/ensure-main-project.sh
Comment (spiffxp) [inline infra/gcp/ensure-main-project.sh:88]: /hold
The rest of this I would like to wait to merge until I land a refactor of this file (https://github.com/kubernetes/k8s.io/pull/1974)

I am happy to take up redoing this part to fit in with the refactor as followup, or leaving it to you to ensure someone other than me understands the structure I tried to put in place

**Review by spiffxp — CHANGES_REQUESTED** [review]

File: infra/gcp/ensure-main-project.sh
Comment (ameukam) [reply infra/gcp/ensure-main-project.sh:62]: > "artifacts" clashes with the sig-release meaning of "binaries and container images that represent a given release of a given subproject"

The buckets are prefixed (`k8s-infra-prow`) with a reference. I don't think we should have confusion with other projects. 

Let's go with `k8s-infra-prow-results`, it's more aligned with the needs of this instance.

File: infra/gcp/ensure-main-project.sh
Comment (ameukam) [reply infra/gcp/ensure-main-project.sh]: @spiffxp made a refactor now #1974 is merged.

File: infra/gcp/ensure-main-project.sh
Comment (spiffxp) [inline infra/gcp/ensure-main-project.sh:185]: I think you are going to want to reuse the existing service account(s) that prow jobs run as instead of creating a new service account

**Review by spiffxp — CHANGES_REQUESTED** [review]
I'm not sure provisioning new service accounts is required here. For the existing prow build clusters their service accounts are setup via terraform

File: infra/gcp/ensure-main-project.sh
Comment (spiffxp) [reply infra/gcp/ensure-main-project.sh:185]: So tl;dr I would create a `k8s-infra-prow@kubernetes-public` service account for the prow control plane, since it may require more than just write access to this one bucket.

Since you're doing this for the cluster in kubernetes-public, you could do all this in https://github.com/kubernetes/k8s.io/blob/main/infra/gcp/clusters/projects/kubernetes-public/aaa

But no objections to iterating in bash if you need to.

File: infra/gcp/ensure-main-project.sh
Comment (spiffxp) [inline infra/gcp/ensure-main-project.sh:213]: This is basically https://github.com/kubernetes/k8s.io/blob/4d98e1c294c6e6158eca1efb9e6a4f8c3003a572/infra/gcp/ensure-conformance-storage.sh#L104-L131

You could parameterize it on project and pull into lib_iam.sh for re-use.  Can be done as followup though

File: infra/gcp/ensure-main-project.sh
Comment (ameukam) [reply infra/gcp/ensure-main-project.sh:185]: @spiffxp I'll add a second commit with a HCL version.

Comment (ameukam) [reply infra/gcp/ensure-main-project.sh:185]: @spiffxp The Terraform version is done in [`03d55ef`](https://github.com/kubernetes/k8s.io/pull/1909/commits/03d55efa685fda4718553202c171b378029223dc).

File: infra/gcp/lib_gsm.sh
Comment (spiffxp) [inline infra/gcp/lib_gsm.sh:111]: merge conflict, you want to keep HEAD

**Review by spiffxp — CHANGES_REQUESTED** [review]

**Comment by spiffxp** [comment]
/close
closing this in favor of https://github.com/kubernetes/k8s.io/pull/2181

**Pull request closed without merging**
